controllers/gitrepository_controller.go

```diff
@@ -725,59 +732,52 @@ func (r *GitRepositoryReconciler) reconcileInclude(ctx context.Context,
 func (r *GitRepositoryReconciler) gitCheckout(ctx context.Context,
 	obj *sourcev1.GitRepository, authOpts *git.AuthOptions, dir string, optimized bool) (*git.Commit, error) {
 	// Configure checkout strategy.
-	checkoutOpts := git.CheckoutOptions{RecurseSubmodules: obj.Spec.RecurseSubmodules}
+	cloneOpts := git.CloneOptions{
+		RecurseSubmodules: obj.Spec.RecurseSubmodules,
+		ShallowClone:      true,
+	}
 	if ref := obj.Spec.Reference; ref != nil {
-		checkoutOpts.Branch = ref.Branch
-		checkoutOpts.Commit = ref.Commit
-		checkoutOpts.Tag = ref.Tag
-		checkoutOpts.SemVer = ref.SemVer
+		cloneOpts.Branch = ref.Branch
+		cloneOpts.Commit = ref.Commit
+		cloneOpts.Tag = ref.Tag
+		cloneOpts.SemVer = ref.SemVer
 	}
 
 	// Only if the object has an existing artifact in storage, attempt to
 	// short-circuit clone operation. reconcileStorage has already verified
 	// that the artifact exists.
 	if optimized && conditions.IsTrue(obj, sourcev1.ArtifactInStorageCondition) {
 		if artifact := obj.GetArtifact(); artifact != nil {
-			checkoutOpts.LastRevision = artifact.Revision
+			cloneOpts.LastObservedCommit = artifact.Revision
 		}
 	}
 
 	gitCtx, cancel := context.WithTimeout(ctx, obj.Spec.Timeout.Duration)
 	defer cancel()
 
-	checkoutStrategy, err := strategy.CheckoutStrategyForImplementation(gitCtx,
-		git.Implementation(obj.Spec.GitImplementation), checkoutOpts)
+	var gitReader git.RepositoryReader
+	var err error
+
+	switch obj.Spec.GitImplementation {
+	case sourcev1.LibGit2Implementation:
+		gitReader, err = libgit2.NewClient(dir, authOpts)
+	case sourcev1.GoGitImplementation:
+		gitReader, err = gogit.NewClient(dir, authOpts)
+	default:
+		err = fmt.Errorf("invalid Git implementation: %s", obj.Spec.GitImplementation)
+	}
 	if err != nil {
 		// Do not return err as recovery without changes is impossible.
-		e := &serror.Stalling{
-			Err:    fmt.Errorf("failed to configure checkout strategy for Git implementation '%s': %w", obj.Spec.GitImplementation, err),
-			Reason: sourcev1.GitOperationFailedReason,
-		}
+		e := serror.NewStalling(
```

Comment:
While reviewing https://github.com/fluxcd/source-controller/pull/945, I noticed that this stalling error may not be appropriate for the above operation. The old code had stalling error because `CheckoutStrategyForImplementation()` failure actually meant that retrying will not make any change. But for the above, new git client creation can fail due to various reasons and a subsequent retry may succeed.
It may be better to make sure it's an invalid git implementation error before returning stalling error. For other error while creating the git client, a generic error which results in a retry may be better.